### PR TITLE
Zerocopy into target files

### PR DESCRIPTION
**Pull request by dhanji**
This is more efficient than copying into a buffer and then into the output file. The new write() method that takes a size argument is needed for this to work. 

The copy operation only transfers between DMA buffers and avoids the CPU altogether. This should produce:
- lower CPU and memory overhead during mvn downloads
- more predictable file write times as they are not affected by thread scheduling
- faster transfers between files with the file system cache

Note: Need to test against platforms that do not support zerocopy to ensure graceful degrade.

### CODE REVIEW

**Comment by ifedorenko** [comment]
I'd use try-with-resource... although we should probably consider rewriting the whole thing.

File: src/test/java/io/takari/filemanager/DefaultFileManagerTest.java [hunk 1]
```diff
@@ -386,7 +387,7 @@ public void testMultipleManagerInstancesShareTheSameLockTable() throws Exception
   @Test
   public void testCopiedFilesHaveTheSameLastModifiedTime() throws Exception {
     File source = TestFileUtils.createTempFile("bytes");
-    source.setLastModified(source.lastModified() - 86400000); // -1 day
+    assert source.setLastModified(source.lastModified() - TimeUnit.DAYS.toMillis(1)); // -1 day
```
Comment (ifedorenko) [inline src/test/java/io/takari/filemanager/DefaultFileManagerTest.java]: out of curiosity, why did you use assertion statement and not junit `assertTrue`? I think surefire enables assertions by default, but eclipse, for example, does not. junit Assert can't be disabled and seems more appropriate for unit tests.

Comment (dhanji) [inline src/test/java/io/takari/filemanager/DefaultFileManagerTest.java]: It's just to satisfy (and capture) the invariant that setLastModified() return true--I never expect it to fail but if it does something serious is wrong (like the OS write failed) which is not really a test failure.

**Comment by dhanji** [comment]
Yea I normally use try-with-resource but wasn't sure if you need to support Java 6 or something. I can change it.

**Pull request closed without merging**